streaming/src/test/scala/org/apache/spark/streaming/scheduler/RateControllerSuite.scala

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.scheduler
+
+import scala.collection.mutable
+import scala.reflect.ClassTag
+import scala.util.control.NonFatal
+
+import org.scalatest.Matchers._
+import org.scalatest.concurrent.Eventually._
+import org.scalatest.time.SpanSugar._
+
+import org.apache.spark.streaming._
+import org.apache.spark.streaming.scheduler.rate.RateEstimator
+
+class RateControllerSuite extends TestSuiteBase {
+
+  override def actuallyWait: Boolean = true
+
+  test("rate controller publishes updates") {
+    val ssc = new StreamingContext(conf, batchDuration)
+    val dstream = new MockRateLimitDStream(ssc, Seq(Seq(1)), 1)
+    val output = new TestOutputStreamWithPartitions(dstream)
```

Comment:
Tests that update the rate and checks whether receiver gets the update should not require running the batches. 

If you need runStreams just to "wait" the batches run, then it is cleaner to just use normal clock (no manual clock) and then just use eventually to wait for the necessary condition to be fulfilled. Now, in some places, you are calling runstreams, as well as using eventually.  This change would simplify it further 

```
ssc.start()

eventualy(...) {
}
```
Regarding stopping it twice, if you are anyways adding code in a `try..catch` or `after` to stop the context, might as well use just `usingStreamingContext` or just `after/afterAll` forget about the `TestSuiteBased` and `runStreams`